gufe/storage/resultsclient.py

```diff
@@ -0,0 +1,141 @@
+from typing import Any
+
+from gufe.storage.resultstore import ResultStore
+from gufe.storage.metadatastore import JSONMetadataStore
+
+
+class _ResultContainer:
+    """
+    Abstract class, represents all data under some level of the heirarchy.
+    """
+    def __init__(self, parent, path_component):
+        self.parent = parent
+        self._path_component = self._to_path_component(path_component)
+        self._cache = {}
```

Comment:
The "subdirectories" are cached. The use case I was thinking was: Say you get a bunch of paths as strings. You want to iterate over then and do something with each. Rather than instantiate another object for each "subdirectory" (and possibly have to worry about memory cleanup) we ensure uniqueness of each "directory" object by caching it and searching the cache first. This should *not* cache file handlers/contexts returned for the direction.

As for when it happens in the `__getitem__`.

I thought I had some use case where this was potentially a significant issue, although honestly I'm not sure now. It might be a YAGNI.